File: bfs.java
import java.io.*;
import java.util.*;

class Graph {
  private HashMap<Integer, Node> nodeLookup = new HashMap<Integer, Node>();

  class Node {
      // it doesnt matter if the following variables are private
      // cos the outer class can access them regardless
      private int value;
      List<Node> adjacent = new ArrayList<Node>();

      private Node(int value) {
        this.value = value;
      }
  }

  public void addNode(int value) {
    nodeLookup.put(value, new Node(value));
  }

  private Node getNode(int value) {
    return nodeLookup.get(value);
  }

  public void connectNodes(int s, int d) {
    Node source = getNode(s);
    Node destination = getNode(d);
    source.adjacent.add(destination);
    // destination.add(source);
  }

  public Boolean checkPath(int s, int d) {
    return bfs(getNode(s), getNode(d));
  }

  private Boolean bfs(Node s, Node d) {
    LinkedList<Node> nextToVisit = new LinkedList<Node>();
    HashSet<Integer> visited = new HashSet<Integer>();

    nextToVisit.add(s);

    while (!nextToVisit.isEmpty()) {
      Node current = nextToVisit.poll();

      if (visited.contains(current.value)) {
        continue;
      }

      if (current == d) {
        return true;
      }

      visited.add(current.value);

      for (Node n: current.adjacent) {
        nextToVisit.add(n);
      }
    }

    return false;
  }
}

public class bfs {
  public static void main(String[] args) {
    Graph graph = new Graph();
    graph.addNode(1);
    graph.addNode(2);
    graph.addNode(3);
    graph.addNode(4);
    graph.addNode(5);
    graph.connectNodes(1, 2);
    graph.connectNodes(1, 3);
    graph.connectNodes(3, 4);
    System.out.println(graph.checkPath(1, 4));
  }
}

File: bfs.py
class Node():
    def __init__(self, value):
        self.value = value
        self.adjacent = []

    def __repr__(self):
        return("{}".format(self.value))

class Graph():
    def __init__(self):
        self.node_lookup = {}
        # maps integer value to node

    def add_node(self, value):
        self.node_lookup[value] = Node(value)

    def get_node(self, value):
        return self.node_lookup[value]

    def connect_nodes(self, source, destination):
        one = self.get_node(source)
        two = self.get_node(destination)
        one.adjacent.append(two)
        # two.adjacent.append(one)
        #^ OPTIONAL?? children cant find their parents without this...

    def check_path(self, source, destination):
        # ^parameters are integers
        return self.bfs(self.get_node(source), self.get_node(destination))


    def bfs(self, source, destination):
        # ^parameters are nodes

        next_to_visit = []
        visited = set()

        next_to_visit.append(source)

        while next_to_visit:
            current = next_to_visit.pop(0)

            if current.value in visited:
                continue

            if current is destination:
                return True

            visited.add(current.value)

            for node in current.adjacent:
                next_to_visit.append(node)

        return False

if __name__ == "__main__":
    graph = Graph()
    graph.add_node(1)
    graph.add_node(2)
    graph.add_node(3)
    graph.add_node(4)
    graph.add_node(5)
    graph.connect_nodes(1, 2)
    graph.connect_nodes(1, 3)
    graph.connect_nodes(3, 4)
    print(graph.check_path(1, 4))
